### PR TITLE
Optimizations to training

### PR DESCRIPTION
**Pull request by orlink**
Optmizations to training:
- For accuracy and flexibility: 
    - added a tokenize_min_batch_size parameter to model_trainer, in order to avoid overfitting - always predicting end of sentence mark at the end of the input text. This was due to the multitude of small .tsv files / tokenized documents, all ending on an end of sentence mark. The new algorithm batches training documents until the total tokens exceed this new parameter before tokenizing the batch. The optimal value of this parameter depends on the number of languages and training epochs as well as the requirements of the application. For 6 language and 2 epochs a value of 40k may work, while for 22 languages and 1 training epoch, a value of 120k may be more appropriate.
   - parameterized the stride. By the way, 150 seems to work better than 100.  
   - multiple data archives can be specified via the 'data' list parameter of the model_trainer class (the 'data' property in model_final_suite_task2.json)

- For speed: paralellized data loading by language

- For memory: continuous removal of already tokenized data from memory, reducing the required memory for training by half or more.

### CODE REVIEW

**Comment by oliverguhr** [comment]
Thank you @orlink. It's very kind that you share your improvements with us.

**Comment by orlink** [comment]
My pleasure. Still waiting on the first training run to finish to see what effect that second to last commit - 876eff4 - about the question marks,  will have.

**Comment by orlink** [comment]
Yeah, I can report now that the new model improved greatly in recognizing questions. batches_ending_on_question_mark_percent was set to 75. Next, I am thinking I may try with variable stride so as to have a bit more samples that start with whole sentences... But I am training on an A100 GPU so each experiment takes more than 6 hours - I have to guess well beforehand. 

Oliver, do you have any idea why the models, including your original xlm-large one, struggle particularly with questions in German (I don't know the language)? For example these:
"Was treibt dich hierher",
"Suchst du nur Sex",
"Na Toni alles gut bei dir",
"Hast du kein Selbstvertrauen Moritz",
"kann ich das auch in  also jetzt sage ich mal zwei",
"Hättest du vielleicht sogar eine richtig große Spielwiese auf der man sich so richtig schön austoben könnte"

--- 
This was a 22 language model with min batch size 120,000. The effect of the batch size is that batches for English or German were reduced from ~3000 to ~1000, which prevents it from overfitting on EOS marks (for that number of languages). 

Some back of the envelope calculations I used, with stride = 150 and English:
33 M words
1.4 M sentences
97% ending on a full stop
3% ending on a question mark
3000 .tsv files (each ending up as a batch in the orginal model_trainer)
1% of them ending on a question mark
With stride 150:
4K + 3K (3K from batch endings) samples ending on a full stop.
0.13K + 0.01K samples ending on a question mark.

Latest model, per language (English):
~ 1000 batches
4K + 0.25 K samples ending on a full stop.
0.13K + 0.75K samples ending on a quesiton mark.

**Comment by oliverguhr** [comment]
I tested these sentences with [the orginal model](https://huggingface.co/oliverguhr/fullstop-punctuation-multilang-large?text=Hast+du+kein+Selbstvertrauen+Moritz).

Here are the results:

"Was treibt dich hierher", -> correct
"Suchst du nur Sex", -> incorrect
"Na Toni alles gut bei dir", -> incorrect
"Hast du kein Selbstvertrauen Moritz", -> correct
"kann ich das auch in also jetzt sage ich mal zwei", -> not a valid question / sentence ()
"Hättest du vielleicht sogar eine richtig große Spielwiese auf der man sich so richtig schön austoben könnte" -> incorrect


The problem might be that this differs a lot from the political speeches, that the model was trained on. These sentences sound like informal language from a dating app or so. If this is not covered in the training data, the model might struggle. 

Also, question marks are the least frequent class of all, so this is could be a balancing problem as well. Question marks are so infrequent that periods are a "safer bet" for the model. Here is the distribution from the original dataset: 

grafik
